### PR TITLE
[WFLY-10522] Support Eclipse MicroProfile Config

### PR DESCRIPTION
**Pull request by jmesnil**
Add the microprofile-config-smallrye subsystem that provides support for
the Eclipse MicroProfile Config using the SmallRye implementation at
https://github.com/smallrye/smallrye-config.

The integrated version is 1.3.0.Final that supports MicroProfile Config
1.3.

The name of the subsystem is `microprofile-config-smallrye`.
It is added by default to *all* WildFly configuration (standalone and
domain).

JIRA: https://issues.jboss.org/browse/WFLY-10522

---

Please make sure your PR meets the following requirements:
- [x] Pull Request title is properly formatted: `[WFLY-XYZ] Subject` or `WFLY-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue(s)
- [x] Pull Request contains description of the issue(s)
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Attached commits represent units of work and are properly formatted

For bigger changes, major and minor component upgrades make sure your PR also meets following requirements:
- [x] Pull Request requires a change to the documentation
- [x] Documentation have been updated accordingly
=> the PR contains the subsystem documentation in the admin guide
- [x] Tests were added to cover changes
=> integration tests checks that deployments can use the MicroProfile Config API
- [x] Analysis was done in wildfly/wildfly-proposals#79
- [x] Test Plan has been done
- [x] Tests were verified in advance

### CODE REVIEW

**Comment by jmesnil** [comment]
- [x] Analysis was done in https://github.com/wildfly/wildfly-proposals/pull/79

**Comment by jmesnil** [comment]
retest this please

**Comment by jmesnil** [comment]
The failures are related to the PR.
* I'm working on fixing the mixed-domain test
* the smallrye-config implementation seems to have issue when the security manager is enabled (it's not using privileged action to get the TCCL).

**Comment by jmesnil** [comment]
retest this please

**Comment by marekkopecky** [comment]
Is it necessary to reload server after one new property is added?

```
[standalone@localhost:9990 /] /subsystem=microprofile-config-smallrye/config-source=aa:add()
{"outcome" => "success"}
[standalone@localhost:9990 /] /subsystem=microprofile-config-smallrye/config-source=aa:write-attribute(name=properties,value={"a"="b","c"="d"})
{
    "outcome" => "success",
    "response-headers" => {
        "operation-requires-reload" => true,
        "process-state" => "reload-required"
    }
}
[standalone@localhost:9990 /] reload 
[standalone@localhost:9990 /] /subsystem=microprofile-config-smallrye/config-source=aa:map-put(name=properties,key=e,value=e)
{
    "outcome" => "success",
    "response-headers" => {
        "operation-requires-reload" => true,
        "process-state" => "reload-required"
    }
}
[standalone@localhost:9990 /] 
```

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+----
+<subsystem xmlns="urn:wildfly:microprofile-config-smallrye:1.0">
+    <config-source name="props">
+        <property name="prop1" value="foo"/>
+        <property name="prop2" value="bar"/>
+    </config-source>
+</subsystem>
+----
+
+=== ConfigSource from Directory
```
Comment (marekkopecky) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc]: It seems that it is also possible to define ConfigSource by "class" name, right? Can you please describe it in this documentation?
```
/subsystem=microprofile-config-smallrye/config-source=aaaa:add(class={
```

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+int numMax; // <2>
+----
+<1> will be set to 5
+<2> will be set to 100
+
+[NOTE]
+This corresponds to the layout used by https://docs.openshift.com/enterprise/3.2/dev_guide/configmaps.html[OpenShift ConfigMaps].
+The `dir` value corresponds to the `mountPath` in the ConfigMap definition in OpenShift or Kubernetes.
+
+
```
Comment (marekkopecky) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc]: Can you please describe also `/subsystem=microprofile-config-smallrye/config-source-provider=` settings in documentation?

**Comment by jmesnil** [comment]
@marekkopecky yes,the server has to be reloaded when properties are modified in the config-source resources for the time being.

The smallrye-config is copying the properties when an application is deployed. Updating the properties in the resource will not update the ConfigSource used by the deployments.

There are enhancements planned for next MicroProfile Config specification to be more dynamic so that ConfigSource can notify their content has changed. When this will be implemented in smallrye-config, reloading the server will no longer be required.

**Comment by wolfc** [comment]
Why is this not coming through a new feature pack provisioned by Galleon?

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc (hunk 1 above)
Comment (jmesnil) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc]: done

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc (hunk 2 above)
Comment (jmesnil) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc]: done

**Comment by jmesnil** [comment]
This new extension depends on the weld (and ee) extensions that are provided by WildFly main feature-pack.

In order to be provided by a separate feature pack, it would first be required to provide separate feature pack for Weld (some kind of baseline for microprofile spec). This goes beyond the scope of this RFE.

**Comment by jmesnil** [comment]
I have fixed the failures that were related to the PR. I don't expect anymore change before it can be reviewed.

File: testsuite/integration/basic/src/test/java/org/wildfly/test/integration/microprofile/config/smallrye/AssertUtils.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.junit.Assert;
+
+/**
+ * @author <a href="http://jmesnil.net/">Jeff Mesnil</a> (c) 2017 Red Hat inc.
+ */
+public class AssertUtils {
+
+    public static void assertTextContainsProperty(String text, String propName, Object propValue) {
+        Assert.assertTrue("Property " + propName + " not found in " + text,
```
Comment (jstourac) [inline testsuite/integration/basic/src/test/java/org/wildfly/test/integration/microprofile/config/smallrye/AssertUtils.java]: Maybe wrap `text` variable into some quotes so it's more easily distinguishable where compared text starts?
```
"Property " + propName + " not found in '" + text + "'"
```

**Comment by jstourac** [comment]
@jmesnil would you mind rebasing this agains master? We would like to add MR for you with some other tests. It would be nice to avoid any possible discrepancies that could arrive in testsuite changes in the meantime. Thank you.

File: microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigSourceDefinition.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                    .build(),
+            create(MODULE, ModelType.STRING, false)
+                    .setAllowExpression(false)
+                    .build())
+            .setAlternatives("properties", "dir")
+            .setAllowNull(true)
+            .setAttributeMarshaller(AttributeMarshaller.ATTRIBUTE_OBJECT)
+            .setRestartAllServices()
+            .build();
+    static AttributeDefinition DIR = SimpleAttributeDefinitionBuilder.create("dir", ModelType.STRING)
```
Comment (bstansberry) [inline microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigSourceDefinition.java]: Filesystem path attributes should be paired with a 'relative-to'.

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+Support for https://microprofile.io/project/eclipse/microprofile-config[Eclipse MicroProfile Config] is provided by
+ the _microprofile-config-smallrye_ subsystem.
+
+[[required-extension]]
+== Required Extension
+
+This extension is included in the standard configurations included in the
+WildFly distribution.
+
+You can also add the extension to a configuration without it either by adding
```
Comment (rsvoboda) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc]: So MP config won't be enabled by default ?

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+@ConfigProperty(name = "num.max")
+int numMax; // <2>
+----
+<1> will be set to 5
+<2> will be set to 100
+
+[NOTE]
+This corresponds to the layout used by https://docs.openshift.com/enterprise/3.2/dev_guide/configmaps.html[OpenShift ConfigMaps].
+The `dir` value corresponds to the `mountPath` in the ConfigMap definition in OpenShift or Kubernetes.
+
```
Comment (rsvoboda) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc]: Please add note that deployment must have CDI enabled - e.g. by META-INF/beans.xml or CDI bean defining annotation.

**Comment by jmesnil** [comment]
@jstourac I've rebased

**Comment by mjurc** [comment]
@jmesnil The last rebase made `wildfly` unbuildable:
`[ERROR] Failed to execute goal on project wildfly-galleon-pack: Could not resolve dependencies for project org.wildfly:wildfly-galleon-pack:pom:14.0.0.Beta2-SNAPSHOT: Failure to find org.wildfly:wildfly-microprofile-config-smallrye:jar:14.0.0.Beta2-SNAPSHOT in http://repository.jboss.org/nexus/content/groups/public/ was cached in the local repository, resolution will not be reattempted until the update interval of jboss-public-repository-group has elapsed or updates are forced -> [Help 1]
`

File: microprofile/config-smallrye/pom.xml [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.wildfly</groupId>
+        <artifactId>wildfly-parent</artifactId>
+        <!--
+        Maintain separation between the artifact id and the version to help prevent
+        merge conflicts between commits changing the GA and those changing the V.
+        -->
+        <version>14.0.0.Beta1-SNAPSHOT</version>
```
Comment (mjurc) [inline microprofile/config-smallrye/pom.xml]: This needs to be bumped to `14.0.0.Beta2-SNAPSHOT`, else `wildfly` is unbuildable

File: microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigProviderService.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2017, Red Hat, Inc., and individual contributors
```
[the remark this answers is not in the document]
Comment (jmesnil) [reply microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigProviderService.java]: Most of the code was taken from https://github.com/wildfly-extras/wildfly-microprofile-config/tree/1.x/extension. I preserved their file headers (including the copyright date)

File: microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigSourceDefinition.java (hunk 4 above)
Comment (jmesnil) [reply microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigSourceDefinition.java]: @bstansberry I have fixed it in a dedicated commit (and check that file system completion works)

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc (hunk 5 above)
Comment (jmesnil) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc]: I mention just above that 
> This extension is included in the standard configurations included in the WildFly distribution

Is the wording ambiguous?

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
```
Comment (tremes) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc]: This seems to me a little bit poor description regarding OpenShift integration. So does this mean that only properties configured in ConfigMap will be supported?

Env. variables defined e.g in Pod or DeploymentConfig will be ignored?

**Comment by jmesnil** [comment]
retest this please

File: microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    @Override
+    public void undeploy(DeploymentUnit context) {
+    }
+
+    private void addDependencies(DeploymentUnit deploymentUnit) {
+        final ModuleSpecification moduleSpecification = deploymentUnit.getAttachment(Attachments.MODULE_SPECIFICATION);
+        final ModuleLoader moduleLoader = Module.getBootModuleLoader();
+
+        moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, MICROPROFILE_CONFIG_API, false, false, true, false));
+        moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, MICROPROFILE_CONFIG_EXTENSION, false, false, true, false));
```
Comment (bstansberry) [inline microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java]: Exposing a subsystem module to a deployment isn't good. We should get clean separation between deployment code and management code, e.g. so we can do things like https://docs.google.com/document/d/1OK8izSS9DVSOAqXjadG7piIjWOJ4KU2eKQ548ZL26aM/edit#heading=h.n28b42dwjj5x

File: feature-pack/src/main/resources/configuration/domain/template.xml [hunk 11]
```diff
@@ -96,20 +97,31 @@
...
         </host-exclude>
         <host-exclude name="WildFly12.0">
             <host-release id="WildFly12.0"/>
             <excluded-extensions>
                 <extension module="org.wildfly.extension.ee-security"/>
+                <extension module="org.wildfly.extension.microprofile.config-smallrye" />
             </excluded-extensions>
         </host-exclude>
+        <!--
+        <host-exclude name="WildFly13.0">
```
Comment (bstansberry) [inline feature-pack/src/main/resources/configuration/domain/template.xml]: https://github.com/wildfly/wildfly-core/pull/3416

File: microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java (hunk 10 above)
Comment (jmesnil) [reply microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java]: fixed. the deployments no longer has any dependency on the extension module.

**Comment by jmesnil** [comment]
retest this please
retest this please

**Comment by jmesnil** [comment]
retest this please

**Comment by jmesnil** [comment]
The PR has been updated to integrate tests written by @mjurc (thanks!)

@wildfly/mergers-full I've removed the missing-reqs flag. 

It is ready to be reviewed & merged as it as has addressed all reported issues or requested changes.

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc (hunk 6 above)
Comment (jmesnil) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc]: I have updated the doc

File: feature-pack/src/main/resources/configuration/domain/template.xml (hunk 11 above)
Comment (jmesnil) [reply feature-pack/src/main/resources/configuration/domain/template.xml]: I will uncomment that in a subsequent PR as there is not yet a Core release with that key

File: microprofile/config-smallrye/pom.xml [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+            <groupId>${project.groupId}</groupId>
+            <artifactId>wildfly-undertow</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>wildfly-weld</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.jboss.logging</groupId>
+            <artifactId>jboss-logging-annotations</artifactId>
```
Comment (jamezp) [inline microprofile/config-smallrye/pom.xml]: This should also include:
```
<!-- This is a compile-time dependency of this project, but is not needed at compile or runtime by other
      projects that depend on this project.-->
<scope>provided</scope>
<optional>true</optional>
```

File: microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigSourceDefinition.java [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+                    .setAllowExpression(false)
+                    .build(),
+            create(MODULE, ModelType.STRING, false)
+                    .setAllowExpression(false)
+                    .build())
+            .setAlternatives("properties", "dir")
+            .setAllowNull(true)
+            .setAttributeMarshaller(AttributeMarshaller.ATTRIBUTE_OBJECT)
+            .setRestartAllServices()
+            .setCapabilityReference("org.wildfly.management.path-manager")
```
Comment (jamezp) [inline microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigSourceDefinition.java]: Should this reference be here or on the `DIR` attribute?

File: microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+import org.jboss.modules.Module;
+import org.jboss.modules.ModuleIdentifier;
+import org.jboss.modules.ModuleLoader;
+
+/**
+ * Add dependencies required by deployment unit to access the Config API (programmatically or using CDI).
+ */
+public class DependencyProcessor implements DeploymentUnitProcessor {
+
+    Logger log = Logger.getLogger(SubsystemDeploymentProcessor.class);
```
Comment (jamezp) [inline microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java]: Is this used anywhere? I'm not seeing it used unless I'm missing something. Having it be named `SubsystemDeploymentProcessor.class` seems a bit odd too.

File: microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/SubsystemDeploymentProcessor.java [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+     * See {@link Phase} for a description of the different phases
+     */
+    public static final Phase PHASE = Phase.POST_MODULE;
+
+    /**
+     * The relative order of this processor within the {@link #PHASE}.
+     * The current number is large enough for it to happen after all
+     * the standard deployment unit processors that come with JBoss AS.
+     */
+    public static final int PRIORITY = 0x4000;
```
Comment (jamezp) [inline microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/SubsystemDeploymentProcessor.java]: Are we no longer putting these in the `Phase`?

File: microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+     * See {@link Phase} for a description of the different phases
+     */
+    public static final Phase PHASE = Phase.DEPENDENCIES;
+
+    /**
+     * The relative order of this processor within the {@link #PHASE}.
+     * The current number is large enough for it to happen after all
+     * the standard deployment unit processors that come with JBoss AS.
+     */
+    public static final int PRIORITY = 0x4000;
```
Comment (jamezp) [inline microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java]: Are we no longer putting these in the `Phase`?

**Review by jamezp — CHANGES_REQUESTED** [review]
It's probably not a big deal, but there are some test classes that are missing the copyright headers.

File: microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/SubsystemDeploymentProcessor.java (hunk 15 above)
Comment (jmesnil) [reply microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/SubsystemDeploymentProcessor.java]: @jamezp Good catch, I set it like that when the code was in wildfly-extras repo and was not "officially" in WildFly code base.

Is it acceptable to use existing priority + offset (e.g. `Phase.POST_MODULE_WELD_PORTABLE_EXTENSIONS + 10`) ?

File: microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java (hunk 16 above)
Comment (jmesnil) [reply microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java]: this one should simply be Phase.DEPENDENCIES_MODULE

File: microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigSourceDefinition.java (hunk 13 above)
Comment (jmesnil) [reply microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigSourceDefinition.java]: fixed

File: microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java (hunk 14 above)
Comment (jmesnil) [reply microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java]: removed

File: microprofile/config-smallrye/pom.xml (hunk 12 above)
Comment (jmesnil) [reply microprofile/config-smallrye/pom.xml]: fixed

**Comment by jmesnil** [comment]
@jamezp I also added the missing copyright headers.

The last thing that may required a change is the priority of the processor. Are you ok with using existing priority + offset (`Phase.POST_MODULE_WELD_PORTABLE_EXTENSIONS + 10`) or must I add an explicit priority to the Phase in wildfly-core (in that case, this PR must be hold until there is a core release with that new Phase).

**Comment by jmesnil** [comment]
retest this please

**Comment by jmesnil** [comment]
@jamezp I've created https://issues.jboss.org/browse/WFCORE-4017 to make sure I don't forget to update the DUP's priority

**Comment by jamezp** [comment]
Perfect thanks @jmesnil. We can make these changes later in the cycle since it's more of a fix than a feature.

**Review by jamezp — APPROVED** [review]
I'm approving this and we can deal with the Phase stuff later.

**Pull request merged**